### PR TITLE
[70_16] Fix with node cause wrong root code and no prog root fallback

### PR DESCRIPTION
**Pull request by UnbSky**
## What

1. With node may cause wrong root code
2. If a prog node is not encountered, the entire document will be rendered, leading to performance issues.

## Why

1. A with node has more than one string child node, and only the last child node is the one that is actually rendered as a string node.
2. When a prog node is not encountered, it uses its own string node for rendering.

### CODE REVIEW

**Comment by da-liii** [comment]
LGTM to me, for me: 
+ Old parser: 2931 ms
+ Tree sitter: 3056 ms

File: src/Plugins/Treesitter/lang_parser.cpp [hunk 1]
```diff
@@ -103,16 +107,20 @@ lang_parser::get_root_node (tree t, int& start_index, int& hash_code) {
...
+    root= t; // Set Root To Self as Fallback
+  }
   hash_code= hash (root);
   get_data_from_root (root, t, start_index);
   return root;
 }
 
 void
 lang_parser::get_data_from_root (tree root, tree line, int& start_index) {
+  if (root->op == with_op) {
```
Comment (jingkaimori) [inline src/Plugins/Treesitter/lang_parser.cpp:121]: 可以利用drd，检测某个宏的子节点是什么类型，比如with宏的子节点是变量名、变量值还是常规内容：

```cpp
root = tree(WITH, "var", "value", tree(CONCAT, "some text", "other things"));
the_drd->get_type_child(root, 0) == TYPE_VARIABLE;
the_drd->get_type_child(root, 1) == TYPE_BINDING;
the_drd->get_type_child(root, 2) == REGULAR;
```
https://github.com/XmacsLabs/moebius/blob/0f5205f27c876fe3a82b4ec3b300db9f56a543d6/moebius/drd/drd_info.hpp#L94

https://github.com/XmacsLabs/moebius/blob/0f5205f27c876fe3a82b4ec3b300db9f56a543d6/moebius/drd/tag_info.hpp#L17-L43

**Review by jingkaimori — COMMENTED** [review]
从With之类的宏中，区分变量和文档内容的语句还能改进：

* 如果只管with，可以用预定义的`WITH`常量；
* 如果需要管其他宏，可以用drd，检测各个子节点的内容类型，区分出不需要高亮的内容

File: src/Plugins/Treesitter/lang_parser.cpp (hunk 1 above)
Comment (UnbSky) [reply src/Plugins/Treesitter/lang_parser.cpp:121]: with(color, red, main)的red类型也是TYPE_REGULAR，只有color是TYPE_VARIABLE,这个和描述好像不太一样

File: src/Plugins/Treesitter/lang_parser.cpp [hunk 2]
```diff
@@ -103,16 +107,20 @@ lang_parser::get_root_node (tree t, int& start_index, int& hash_code) {
       }
     }
   }
-  // if(N(father_ip) == 1){
-  //   cout << "Can not meet Prog " << t << " IP:" << obtain_ip(t) << LF;
-  // }
+  if (N (father_ip) == 1) {
```
[the remark this answers is not in the document]
Comment (UnbSky) [reply src/Plugins/Treesitter/lang_parser.cpp:110]: cpp文件目前是匹配不到prog结点的

File: src/Plugins/Treesitter/lang_parser.cpp (hunk 1 above)
Comment (UnbSky) [reply src/Plugins/Treesitter/lang_parser.cpp]: 使用is_accessible_child接口实现

File: src/Plugins/Treesitter/lang_parser.cpp [hunk 3]
```diff
@@ -131,8 +132,12 @@ lang_parser::get_data_from_root (tree root, tree line, int& start_index) {
     change_line_pos << change_index;
   }
   else {
-    for (tree child_node : root) {
-      get_data_from_root (child_node, line, start_index);
+    for (int i= 0; i < N (root); i++) {
```
Comment (da-liii) [inline src/Plugins/Treesitter/lang_parser.cpp:135]: use root_N instead of N(root)

File: src/Plugins/Treesitter/lang_parser.cpp [hunk 4]
```diff
@@ -170,8 +175,12 @@ lang_parser::get_code_from_root (tree root, string& code, string_u8& code_u8) {
     }
   }
   else {
-    for (tree child_node : root) {
-      get_code_from_root (child_node, code, code_u8);
+    for (int i= 0; i < N (root); i++) {
```
Comment (da-liii) [inline src/Plugins/Treesitter/lang_parser.cpp:178]: use root_N instead of N(root)

**Review by jingkaimori — APPROVED** [review]
LGTM

**Pull request merged**